UI update: possibility to change DSP flow and change DSP settings on the fly

anabolyc opened the pull request:
Hey folks. Please treat this as a draft. To kickoff a work around [persisted UI changes](https://github.com/CarlosDerSeher/snapclient/issues/156) I reviewed current UI and updated it to fully support curent DSP capabilities

- It starts with stereo flow by default (pass-through, no shanges, apart for applying soft volume)
- It expose possible settings to `ui_http_server` component and it get served to UI.
- `ui_http_server` reads and stores DSP settings in the NVS, but keep using SPIFFS for static web files.
- `ui_http_server` sends stored settings to `dsp_processor` upon boot
- Web UI loads current settings, as they are read from the NVS
- When UI change DSP flow or Flow settings, the are posted to `ui_http_server`
- `ui_http_server` persist changes on the NVS and send updates to the `dsp_processor`

To do:
[X] - Edit and persist DSP settings on NVS
[X] - Edit hostname and persist on NVS
[X] - Remove DSP settings from SDKCONFIG, default to no processing at start
[X] - Fix pipeline errors

Before it gets too big to digest, I'd like to hear your opinion

Here are soe screenshots. Default state
<img width="1420" height="501" alt="image" src="https://github.com/user-attachments/assets/1fb58433-79c7-49c5-b12f-01e124b34463" />

Bass/treble flow
<img width="1650" height="788" alt="image" src="https://github.com/user-attachments/assets/c55bfef8-5f5f-4894-8e13-3a9ea150b6ed" />

Bass boost flow
<img width="1650" height="788" alt="image" src="https://github.com/user-attachments/assets/f258c4ea-4288-430a-9384-f523c722b48a" />

Bi-amp flow
<img width="1650" height="788" alt="image" src="https://github.com/user-attachments/assets/c43e4e16-74d9-4672-ae9d-4588545f763a" />

As a next step I think to remove these settings from the SDKCONFIG, as they can be changed on the fly now, woth default fall to no processing.

Small note, it is based on unmerged changes, I will do tidying up (rebase) before wrapping it up.

luar123 commented:
The tests are building with idf v5.5.1, this is why they through errors. In v5.1.x these errors were only warnings.

CarlosDerSeher commented:
Without looking at the code, this seems nice. Great you put in so much effort 👌

anabolyc commented:
> Why did you decide to keep spiffs?

esp-idf built-in packing of files on the filesystem into a spiffs partition, and it is read-only at all times. I thought, why not. Do you recommend another approach?

anabolyc commented:
Here is what I have now:
- hostname is possible to change in the UI (requires a restart to apply)
- it is stored in NVS and used everywhere
- initially value from SDKCONFIG is used

(page 1)
<img width="1650" height="788" alt="image" src="https://github.com/user-attachments/assets/3e637836-8fc7-4412-b06a-61ab6376f546" />
(page 2)
<img width="1649" height="884" alt="image" src="https://github.com/user-attachments/assets/09053c31-e96d-4c8a-8ac0-0dcb60424a5b" />

CarlosDerSeher commented:
I think I am ok with the first solution. How much is a bit increased binary size? I guess there is still plenty left of the ~2MB that we have for the application?

CarlosDerSeher commented:
> esp-idf built-in packing of files on the filesystem into a spiffs partition, and it is read-only at all times. I thought, why not. Do you recommend another approach?

@anabolyc sorry didn't see that.
No I don't just because you wrote it is deprecated which it isn't is it?

And I don't think it is read only...

You could include the html webpage as a string in a c file somewhere.

anabolyc commented:
> That's a very nice addition. I am just wondering if we can simplify the architecture a bit. At the moment we have
> 
> * dsp_processor loads init values directly from dsp_settings
> * http_server gets flow and parameter from dsp_processor and sends new settings to dsp_processor and dsp_settings
> * http_server calls switch_flow on dsp_processor and receives settings
> 
> Some ideas:
> 
> * We could remove the dsp_processor->dsp_settings dependency by adding a function `dsp_all_params_t get_all_params()` function (similar to the `get_json()` function) to dsp_settings and pass the result to `dsp_processor_init()`
> * Bigger change: Store only parameters of current flow in dsp_processor, defaults to Stereoflow in init. Other defaults would go to dsp settings. After init get flow and params from settings and call dsp processors `set_params_for_flow` and `switch_flow` methods. http_server then needs to get the parameters always from dsp_settings and send them to dsp_processor.

Let me explain what I'm trying to do, and you may help me to go along this logic

I want to split the UI logic (`ui_http_server`) from the storage logic (`dsp_processor_settings`) and actual DSP processor logic (`dsp_processor`) to have clear responsibilities of each module with small, clear interfaces. 

While I only need to store and retrieve values, all stays fairly simple: both UI and DSP processor access `dsp_processor_settings`, and changes are applied after reboot - no synchronization needed yet - this the `settings_manager` case, all fairly simple and modular.

When I start updating values on the fly, I need to notify my clients (`dsp_processor`) that the settings have changed. I guess I should have moved the queue logic to `dsp_processor_settings` as well, instead of `ui_http_server`, where it was when I started. What do you think?

I'm starting to implement UI for the TAS5805 DAC, and I'm trying to apply the same approach: I don't want too many DAC specifics in the `ui_http_server`, and I don't want NVS stuff in the `tas5805m` component either. I'm creating `tas5805m_settings` component to bridge those together and add NVS underneath.

luar123 commented:
> I want to split the UI logic (ui_http_server) from the storage logic (dsp_processor_settings) and actual DSP processor logic (dsp_processor) to have clear responsibilities of each module with small, clear interfaces.

That makes a lot of sense.

> I guess I should have moved the queue logic to dsp_processor_settings as well, instead of ui_http_server, where it was when I started. What do you think?

That could also make sense. I am also not sure.

>> * We could remove the dsp_processor->dsp_settings dependency by adding a function dsp_all_params_t get_all_params() function (similar to the get_json() function) to dsp_settings and pass the result to dsp_processor_init()
>> * Bigger change: Store only parameters of current flow in dsp_processor, defaults to Stereoflow in init. Other defaults would go to dsp settings. After init get flow and params from settings and call dsp processors set_params_for_flow and switch_flow methods. http_server then needs to get the parameters always from dsp_settings and send them to dsp_processor.

Basically both ideas are trying to separate the modules even more by removing the call from dsp_processor to dsp_settings. With this change one could remove the settings and http_server_ui modules completely if only the snapclient should be integrated into another project.

The second idea additionally tries to simplify the dsp_processor by keeping unneeded settings in the dsp_settings module. Probably this would be easier and more logical if you would move the queue to dsp_settings, but I am not sure how this would look like.

CarlosDerSeher commented on components/dsp_processor/dsp_processor.c:
```diff
@@ -72,51 +65,66 @@ void dsp_processor_init(void) {
+
+  // Load saved parameters from NVS for all flows
+  ESP_LOGI(TAG, "%s: Loading saved parameters from NVS", __func__);
+  for (int flow = 0; flow < DSP_FLOW_COUNT; flow++) {
```
Why did you choose `flow = 0` instead of using `flow = dspfStereo`?

anabolyc replied:
It is a bit more robust against change of order in enum, altough less readable. I can change, if you prefer `flow = dspfStereo` more

anabolyc commented on components/ui_http_server/CMakeLists.txt:
Hey @CarlosDerSeher 

I've implemented EMBED_TXTFILES instead of SPIFFS
I couldn't find an official announcement of its deprecation. I think it was in the Arduino framework release changes, so it may not be applicable to IDF. But in terms of safety and speed, EMBED_TXTFILES wins anyway.

luar123 commented on components/dsp_processor/dsp_processor.c:
```diff
@@ -712,9 +738,114 @@ int dsp_processor_worker(void *p_pcmChnk, const void *p_scSet) {
+/**
+ * Switch to a different DSP flow
+ */
+esp_err_t dsp_processor_switch_flow(dspFlows_t flow) {
```
Is this still needed? And if not I think there is no need anymore to store parameters of all flows in dsp_processor, right?

anabolyc replied:
removed

luar123 reviewed (commented):
I think that is much cleaner now. But probably you could remove a lot of stuff from dsp_processor now.

anabolyc commented on components/dsp_processor_settings/dsp_processor_settings.c:
```diff
@@ -0,0 +1,642 @@
+        
+        // Notify DSP processor if this is the active flow
+        if (flow == settings_cache.active_flow) {
+            QueueHandle_t queue = dsp_processor_get_filter_queue();
```
You're right, I did the same in the tas5805m_settings module. Much cleaner dependency tree. Will push change after testing

anabolyc commented:
@luar123 please have a look, I've reversed `dsp_processor` dependency on `dsp_processor_settings`, I think I start to like the structure now ;) Also added a restart button on the UI

<img width="876" height="795" alt="image" src="https://github.com/user-attachments/assets/3338f8d4-491e-46ea-8c77-aeb4be21fe51" />

luar123 commented on components/dsp_processor/dsp_processor.c:
```diff
@@ -236,17 +253,42 @@ int dsp_processor_worker(void *p_pcmChnk, const void *p_scSet) {
-  if (xQueueReceive(filterUpdateQHdl, &filterParams, pdMS_TO_TICKS(0)) ==
-      pdTRUE) {
+  // If parameters were changed by API, copy them from centralized storage
+  if (params_changed) {
```
I think you need to guard this by a mutex.
Or maybe better use a mutex for signaling:
`if xSemaphoreTake(update_params_mutex, pdMS_TO_TICKS(0) {...}`
and in `init` you need to initialize and take the mutex and in `dsp_processor_update_filter_params` you give the mutex.

luar123 replied:
binary semaphore might be better than mutex

anabolyc replied:
Changed, please have a look. Not sure I'm skilled enough to see the benefit of the binary semaphore, did the way I can ;)

luar123 replied:
That was not what I meant. Please see my suggestions:

```suggestion
  if (xSemaphoreTake(paramsChangedSemaphoreHandle, 0) == pdTRUE) {
```

luar123 reviewed (commented):
Thanks for all the effort! I think the architecture is much better now. I looked mostly at a higher level (main and dsp_processor, not so much into http_server) and didn't test it yet. So @CarlosDerSeher should also have a closer look.

One thing I still don't understand is why you need the parameters of all flows stored in dsp_processer. But keeping it as is also would be fine for me.

luar123 commented:
> It is there for `dsp_processor` modularity. In case `dsp_processor_switch_flow` is called, `dsp_processor` should be able to supply default parameters, even if `dsp_processor_set_params_for_flow` was never called. I think it is more robust against API misuse this way.

Ok, if you want to keep this API that makes sense.

luar123 commented on components/dsp_processor/dsp_processor.c:
```diff
@@ -11,22 +11,46 @@
 
-static QueueHandle_t filterUpdateQHdl = NULL;
+// Legacy queue removed. Use params_changed flag to notify worker of updates.
+static volatile bool params_changed = false;
```
```suggestion
// Legacy queue removed. Use paramsChangedSemaphoreHandle to notify worker of updates.
static SemaphoreHandle_t paramsChangedSemaphoreHandle = NULL;
```

luar123 reviewed (changes requested):
Tested this on one of my esp32 boards:
- static ip config did not work, see comment with fix
- When opening the dsp settings page for the first time it shows the EQ flow settings, but should be Stereo flow.
- I merged it locally with #172 because otherwise I would get a lot of sync issues (running it on my worse board without psram). So I would suggest to first merge sync_rework branch+#172 and then this PR.
Apart from these it is looking nice.

luar123 commented on components/dsp_processor_settings/dsp_processor_settings.c:
```diff
@@ -0,0 +1,654 @@
+	// persisted configuration after both modules are initialized.
+	// Note: caller (main) must initialize dsp_processor before calling
+	// dsp_settings_init() so these calls succeed.
+	dspFlows_t active = dspfEQBassTreble;
```
Set this to stereo?

luar123 commented on components/dsp_processor_settings/dsp_processor_settings.c:
```diff
@@ -0,0 +1,654 @@
+	}
+
+	// Add active flow
+	dspFlows_t active_flow = dspfEQBassTreble; // default
```
stereo?

anabolyc replied:
This is addressed

luar123 reviewed (commented):
When no nvs settings are present (after `esptool erase-flash`) there is no audio output. See comments for some related issues. I think the problem is that flow parameters are all set to 0 and flow is switched to EQ flow.

anabolyc commented:
> Not sure if this is because I used IDF 5.1.5 but I get lots of these compile errors

> I didn't look into this further now as I don't have the muse to change my dev environment tonight. I am on your branch `features/change-hostname-via-ui`

I only use 5.5.1 in develop. I think @luar123 mentioned somewhere that otherwise 5.1.1 generates a lot of those errors

luar123 commented:
Tried to build it with v5.1.5 and also get `error: format '%d' expects argument of type 'int', but argument 7 has type 'int32_t'`. I guess the settings changed in 5.5.1, but I think it would be good to fix them. @anabolyc you can compile with `-Werror=format=` to see the errors in v5.5.1.

The errors in dsp_processor.c don't make much sense, e.q. `error: unknown type name 'SemaphoreHandle_t'`. This type is also used in player.c

>  I think luar123 mentioned somewhere that otherwise 5.1.1 generates a lot of those errors

It was the other way round: 5.5.1 gave errors when building master (should be fixed by now)

luar123 commented:
> Could be a missing header inclusion in dsp processor?

good point, maybe add `#include "freertos/semphr.h"`

anabolyc commented:
> -Werror=format=

When you build under 5.1.1, how do you pass this error? Isn't it bound to new 5.5.1 API?
```
/Users/andriy/prj/esparagus-media-center/esp32-snapclient-change-hostname-via-ui/components/network_interface/network_interface.c: In function 'network_get_netif_from_desc':
/Users/andriy/prj/esparagus-media-center/esp32-snapclient-change-hostname-via-ui/components/network_interface/network_interface.c:58:10: error: implicit declaration of function 'esp_netif_find_if'; did you mean 'esp_netif_init'? [-Werror=implicit-function-declaration]
   58 |   return esp_netif_find_if(netif_desc_matches_with, (void *)desc);
      |          ^~~~~~~~~~~~~~~~~
      |          esp_netif_init
```

luar123 reviewed (approved):
Perfect, works great!

CarlosDerSeher commented on components/dsp_processor_settings/dsp_processor_settings.c:
```diff
@@ -0,0 +1,654 @@
+ */
+static void make_flow_key(char *out_key, size_t out_size, dspFlows_t flow,
+						  const char *param) {
+	snprintf(out_key, out_size, "flow_%d_%s", (int)flow, param);
```
check for NULL pointers passed

anabolyc replied:
Corrected

CarlosDerSeher commented on components/dsp_processor_settings/dsp_processor_settings.c:
```diff
@@ -0,0 +1,654 @@
+		return ESP_ERR_INVALID_STATE;
+
+	char key[32];
+	make_flow_key(key, sizeof(key), flow, param_name);
```
Check for valid flow before calling this then `make_flow_key()` could stay as it is

anabolyc replied:
Corrected

CarlosDerSeher reviewed (changes requested):
@anabolyc 
Is it really necessary to do the restart of ESP after settings changed? DSP settings are accepted on the fly as I see it? Couldn't you just kill the netconn connection to snapserver when settings are saved and all those other settings will be applied on the reconnect attempt?

Maybe we should make `enum filtertypes` a type of its own instead of using `int` as function parameter type? This would require `ptype_t` to by changed, etc.

anabolyc commented:
> @anabolyc Is it really necessary to do the restart of ESP after settings changed? DSP settings are accepted on the fly as I see it? Couldn't you just kill the netconn connection to snapserver when settings are saved and all those other settings will be applied on the reconnect attempt?
> 
> Maybe we should make `enum filtertypes` a type of its own instead of using `int` as function parameter type? This would require `ptype_t` to by changed, etc.

Thanks for the feedback. I'll try to address these this week.

anabolyc commented on components/dsp_processor/dsp_processor.c:
```diff
@@ -712,9 +770,76 @@ int dsp_processor_worker(void *p_pcmChnk, const void *p_scSet) {
+  }
+  
+  // Update centralized storage for this specific flow
+  if (params_mutex) xSemaphoreTake(params_mutex, portMAX_DELAY);
```
I guess it will go forward without mutex protection. I added warning logs, but what else we can do?

anabolyc commented on components/dsp_processor/dsp_processor.c:
```diff
@@ -712,9 +770,76 @@ int dsp_processor_worker(void *p_pcmChnk, const void *p_scSet) {
+ * Switch to a different DSP flow
+ */
+esp_err_t dsp_processor_switch_flow(dspFlows_t flow) {
+  ESP_LOGI(TAG, "%s: switching from flow %d to %d", __func__, all_params.active_flow, flow);
```
Corrected

anabolyc commented on components/dsp_processor_settings/dsp_processor_settings.c:
```diff
@@ -0,0 +1,654 @@
+/**
+ * Get parameters for a specific flow
+ */
+esp_err_t dsp_settings_get_flow_params(dspFlows_t flow,
```
Corrected. Also removed `fc_2` and `gain_2`. Turned out they were dummies in the original UI, no DSP is actually implementing mid freq gain change

anabolyc commented on components/dsp_processor_settings/dsp_processor_settings.c:
```diff
@@ -0,0 +1,654 @@
+	return ESP_OK;
+}
+
+esp_err_t dsp_settings_save_active_flow(dspFlows_t flow) {
```
Corrected

anabolyc commented:
> @anabolyc Is it really necessary to do the restart of ESP after settings changed? DSP settings are accepted on the fly as I see it? Couldn't you just kill the netconn connection to snapserver when settings are saved and all those other settings will be applied on the reconnect attempt?

DSP settings do not require a restart for sure. For the snapclient part, I'm not sure, I don't have that much experience with it, but I can certainly try if you point me to it.

> Maybe we should make `enum filtertypes` a type of its own instead of using `int` as function parameter type? This would require `ptype_t` to by changed, etc.

To be honest, I tried to avoid changing DSP internals, other than aligning with UI capabilities, since it was not my intention for this PR. But I'm open for suggestions, if you can elaborate little more.

anabolyc commented:
@CarlosDerSeher I moved 'Restart' button to each settings block, for clarity. It would only activate when changes are applied. We can eliminate them one by one, when the underlying implementation would support applying without a restart. But to be honest, I'd be perfectly fine with restarting the device here, as it would not happen often, and it really takes a few seconds to reboot, UI recovers automatically within seconds.

<img width="1265" height="821" alt="image" src="https://github.com/user-attachments/assets/aa51ae20-ef20-487f-9fda-ba45c6bf5125" />

CarlosDerSeher commented:
> For the snapclient part, I'm not sure, I don't have that much experience with it, but I can certainly try if you point me to it.

This was just an idea, not a must have. You could try to close netconn when the netconn read function times out. Just check a flag you set from dsp for quick and dirty test. Maybe you could even use needs restart flag for this.

> To be honest, I tried to avoid changing DSP internals, other than aligning with UI capabilities

This is totally OK with me, it can be done some other time. It just jumped at me when reviewing the changes.

anabolyc commented:
> > For the snapclient part, I'm not sure, I don't have that much experience with it, but I can certainly try if you point me to it.
> 
> This was just an idea, not a must have. You could try to close netconn when the netconn read function times out. Just check a flag you set from dsp for quick and dirty test. Maybe you could even use needs restart flag for this.
> 
> > To be honest, I tried to avoid changing DSP internals, other than aligning with UI capabilities
> 
> This is totally OK with me, it can be done some other time. It just jumped at me when reviewing the changes.

Got it. So let's keep it maybe for the next time, as this PR is already way too heavy.

CarlosDerSeher commented:
@anabolyc so this seems ready to merge. I have one last request though. There are a lot commit messages when squashing. Could you squash this yourself or at least write up a squashed commit messages for me and reduce the content to the essentials?

Commit:
feat: Add UI for DSP settings management and hostname configuration

Major features:
- Added web UI to manage DSP settings and change DSP flow on the fly
- Added UI to change mdns hostname and persist in NVS
- Added restart button to UI with improved placement
- Moved files from SPIFFS to flash embed

Architecture improvements:
- Refactored DSP settings into dedicated component
- Broke dependency loop between ui_http_server and dsp_processor
- Reversed dependency of dsp_processor on dsp_processor_settings
- Fully data-driven front-end implementation

Bug fixes and cleanup:
- Fixed build errors under ESP-IDF v5.1.5
- Fixed default DSP settings producing no sound
- Fixed static snapserver config in sdkconfig
- Added mutex protection to params_change
- Removed unused fc_2 and gain_2 parameters
- Added fail-safe defaults to freq and gain settings
- Code refactoring and cleanup

anabolyc commented:
> @anabolyc so this seems ready to merge. I have one last request though. There are a lot commit messages when squashing. Could you squash this yourself or at least write up a squashed commit messages for me and reduce the content to the essentials?

Done, just waiting for build validation to pass

CarlosDerSeher commented:
> Implemented volume control without lookup tables (-60dB to 0dB range)

This is related to `TAS5805M driver improvements:` isn't it?

anabolyc commented:
> > Implemented volume control without lookup tables (-60dB to 0dB range)
> 
> This is related to `TAS5805M driver improvements:` isn't it?

Right, I didn't pull develop in my fork. I'll update the message in a minute.

Upd: done!

CarlosDerSeher commented:
Not sure what you did now. Could you elaborate?

CarlosDerSeher commented:
Not sure what you did now. Could you elaborate?

anabolyc commented:
> Not sure what you did now. Could you elaborate?

Just updated the commit message. I was collecting changes from my fork's develop, and picked up more than you have here, hence extra lines in the original commit message.

CarlosDerSeher commented:
> - Implemented volume control without lookup tables (-60dB to 0dB range)

So what is this exactly in the code?

anabolyc commented:
> > * Implemented volume control without lookup tables (-60dB to 0dB range)
> 
> So what is this exactly in the code?

It was related to [this change](https://github.com/CarlosDerSeher/snapclient/commit/3528da76846a0f56febfd9e741b2e4e81abecd31)

CarlosDerSeher commented:
But I didn't see it in this PR. Did I just missed it?

anabolyc commented:
> But I didn't see it in this PR. Did I just missed it?

Correct, it is not part of this PR, that's why I removed this part from this PR's commit message

luar123 commented:
> Correct, it is not part of this PR, that's why I removed this part from this PR's commit message

This line is still in ;-)

anabolyc commented:
> > Correct, it is not part of this PR, that's why I removed this part from this PR's commit message
> 
> This line is still in ;-)

Oh, sorry, I was so sure I removed that :) Now it IS removed :)

CarlosDerSeher commented:
@anabolyc after testing your changes thoroughly I found that RAM usage is a bit higher now. I get warnings from `static FLAC__StreamDecoderWriteStatus write_callback()` from this line `pcmData = (uint8_t *)realloc(pcmChunk.outData, pcmChunk.bytes + bytes);` After adding some heap debug I found that there is enough space available but not such a big block. 

```
ESP_LOGW(TAG, "%s, failed to allocate PCM chunk payload (%lu + %u bytes, free heap %u, largest block %u), try again", __func__, 
                                                                                                                            pcmChunk.bytes, 
                                                                                                                            bytes, 
                                                                                                                            heap_caps_get_free_size(MALLOC_CAP_8BIT), 
                                                                                                                            heap_caps_get_largest_free_block(MALLOC_CAP_8BIT));
```

I don't have PSRAM on most of my boards and server is set to 758ms buffer with a sample rate of 44100Hz. Reducing buffer to 700ms makes those warnings go away, Probably I could have reduced it by just one 26ms chunk and it would have been enough but I did not test.

Do you have an idea where the increased RAM usage could come from? I don't have dsp_processor enabled. I didn't see much increase in settings manager.

One idea to mitigate this is to keep `pcmChunk.outData` allocated and just increase it if needed but do not decrease it anymore. Only free it on codec change maybe?

anabolyc commented:
@CarlosDerSeher nut sure, not that experienced with IDF. I think I had to increase few capabilities of the web server to handle increased load, might be also a few json buffers taking extra space. Not sure if that can be moved to PSRAM instead

CarlosDerSeher commented:
No, PSRAM isn't an option as this question is about modules without PSRAM

anabolyc commented:
> No, PSRAM isn't an option as this question is about modules without PSRAM

Maybe a few more things could be put behind compile guards if DSP is not enabled in menuconfig? Probably most of UI can be excluded alltogether, since there will be nothing to configure, isn't it?

CarlosDerSeher commented:
I'd prefer this if possible so we can still build a minimum footprint image.

Only configurable thing would be client name which could also be set through menuconfig.

I'll test this by disabling http server first though

CarlosDerSeher commented:
See my comment in (https://github.com/CarlosDerSeher/snapclient/issues/197#issuecomment-3873805636)
